llvm/tools/sycl-post-link/ModuleSplitter.cpp

```diff
@@ -0,0 +1,433 @@
+
+  return Res;
+}
+
+// TODO: try to move including all passes (cleanup, spec consts, compile time
+// properties) in one place and execute MPM.run() only once.
+void cleanupSplitModule(Module &SplitM) {
+  ModuleAnalysisManager MAM;
+  MAM.registerPass([&] { return PassInstrumentationAnalysis(); });
+  ModulePassManager MPM;
+  // Do cleanup.
+  MPM.addPass(GlobalDCEPass());           // Delete unreachable globals.
+  MPM.addPass(StripDeadDebugInfoPass());  // Remove dead debug info.
+  MPM.addPass(StripDeadPrototypesPass()); // Remove dead func decls.
+  MPM.run(SplitM, MAM);
+}
+
+// The function produces a copy of input LLVM IR module M with only those entry
+// points that are specified in ModuleEntryPoints vector.
+ModuleDesc extractCallGraph(const Module &M,
+                            const EntryPointGroup &ModuleEntryPoints) {
+  SetVector<const GlobalValue *> GVs;
+  collectFunctionsToExtract(GVs, ModuleEntryPoints);
+  collectGlobalVarsToExtract(GVs, M);
+
+  ModuleDesc SplitM = extractSubModule(M, GVs, ModuleEntryPoints);
+  cleanupSplitModule(SplitM.getModule());
+
+  return SplitM;
+}
+
+class ModuleCopier : public ModuleSplitterBase {
+public:
+  using ModuleSplitterBase::ModuleSplitterBase; // to inherit base constructors
+
+  ModuleDesc nextSplit() override {
+    return {releaseInputModule(), nextGroup()};
+  }
+};
+
+class ModuleSplitter : public ModuleSplitterBase {
+public:
+  using ModuleSplitterBase::ModuleSplitterBase; // to inherit base constructors
+
+  ModuleDesc nextSplit() override {
+    return extractCallGraph(getInputModule(), nextGroup());
+  }
+};
+
+} // namespace
+
+namespace llvm {
+
+namespace module_split {
+
+bool ModuleDesc::isEsimd() { return (EntryPoints.GroupId == ESIMD_SCOPE_NAME); }
+
+std::unique_ptr<ModuleSplitterBase>
+getSplitterByKernelType(std::unique_ptr<Module> M, bool SplitEsimd,
+                        bool EmitOnlyKernelsAsEntryPoints) {
+  EntryPointGroupVec Groups = groupEntryPointsByKernelType(
+      *M, SplitEsimd, EmitOnlyKernelsAsEntryPoints);
+  bool DoSplit = (Groups.size() > 1);
+
+  std::unique_ptr<ModuleSplitterBase> Splitter;
+  if (!DoSplit)
+    Splitter.reset(new ModuleCopier(std::move(M), std::move(Groups)));
+  else
+    Splitter.reset(new ModuleSplitter(std::move(M), std::move(Groups)));
+  return Splitter;
+}
+
+std::unique_ptr<ModuleSplitterBase>
+getSplitterByMode(std::unique_ptr<Module> M, IRSplitMode Mode,
+                  bool IROutputOnly, bool EmitOnlyKernelsAsEntryPoints,
+                  bool DeviceGlobals) {
+  EntryPointsGroupScope Scope =
+      selectDeviceCodeGroupScope(*M, Mode, IROutputOnly);
+  EntryPointGroupVec Groups =
+      groupEntryPointsByScope(*M, Scope, EmitOnlyKernelsAsEntryPoints);
+  assert(!Groups.empty() && "At least one group is expected");
+  if (DeviceGlobals)
+    checkImageScopedDeviceGlobals(*M, Groups);
+  bool DoSplit = (Mode != SPLIT_NONE &&
+                  (Groups.size() > 1 || !Groups.cbegin()->Functions.empty()));
```

Comment:
We are checking for at least a single function in the *first* group but why not in *every* one?